Core/include/Acts/Geometry/BlueprintNode.hpp

```diff
@@ -0,0 +1,295 @@
+  /// volume and adds it to this node as a child.
+  /// @param volume The volume to add
+  /// @param callback An optional callback that receives the node as an argument
+  /// @return A reference to the created node
+  StaticBlueprintNode& addStaticVolume(
+      std::unique_ptr<TrackingVolume> volume,
+      const std::function<void(StaticBlueprintNode& cylinder)>& callback = {});
+
+  /// Alternative overload for creating a @ref Acts::StaticBlueprintNode. This
+  /// overload will invoke the constructor of @ref Acts::TrackingVolume and use
+  /// that volume to create the node.
+  /// @param transform The transform of the volume
+  /// @param volumeBounds The bounds of the volume
+  /// @param volumeName The name of the volume
+  /// @param callback An optional callback that receives the node as an argument
+  StaticBlueprintNode& addStaticVolume(
+      const Transform3& transform, std::shared_ptr<VolumeBounds> volumeBounds,
+      const std::string& volumeName = "undefined",
+      const std::function<void(StaticBlueprintNode& cylinder)>& callback = {});
+
+  /// Convenience method for creating a @ref Acts::CylinderContainerBlueprintNode.
+  /// @param name The name of the container node. This name is only reflected
+  ///             in the node tree for debugging, as no extra volumes is created
+  ///             for the container.
+  /// @param direction The direction of the stack configuration. See
+  ///                  @ref Acts::CylinderVolumeStack for details.
+  /// @param callback An optional callback that receives the node as an argument
+  CylinderContainerBlueprintNode& addCylinderContainer(
+      const std::string& name, BinningValue direction,
+      const std::function<void(CylinderContainerBlueprintNode& cylinder)>&
+          callback = {});
+
+  /// Convenience method for creating a @ref Acts::MaterialDesignatorBlueprintNode.
+  /// @param name The name of the material designator node. Used for debugging
+  ///             the node tree only.
+  /// @param callback An optional callback that receives the node as an argument
+  MaterialDesignatorBlueprintNode& addMaterial(
+      const std::string& name,
+      const std::function<void(MaterialDesignatorBlueprintNode& material)>&
+          callback = {});
+
+  /// Convenience method for creating a @ref Acts::LayerBlueprintNode.
+  /// @param name The name of the layer node.
+  /// @param callback An optional callback that receives the node as an argument
+  LayerBlueprintNode& addLayer(
+      const std::string& name,
+      const std::function<void(LayerBlueprintNode& layer)>& callback = {});
+
+  /// @}
+
+  /// Register a @p child to this node.
+  /// @warning This method throws if adding the child would create a
+  ///          cycle in the blueprint tree!
+  /// @param child The child node to add
+  /// @return A reference this node (not the child!)
+  BlueprintNode& addChild(std::shared_ptr<BlueprintNode> child);
```

Comment:
Are we worried that someone might add a child twice? Since a child node does not know it's parent we cannot detect this. Potentially this is just one of the user errors we don't want to deal with

Reply:
I can detect it with the depth I think. The method checks if the depth is 0 and if not will error out. 

Indeed that's a problem, since the building process consumes the tree, i.e. the construction cannot be executed multiple times with the same tree instance.

I'm also checking if there's circles in the tree when adding a child.